jbmc/src/java_bytecode/code_with_references.h

```diff
@@ -33,4 +33,72 @@ struct object_creation_referencet
   optionalt<exprt> array_length;
 };
 
+/// Base class for code which can contain references which can get replaced
+/// before generating actual codet.
+/// Currently only references in array allocations are supported, because this
+/// is currently the only use required by \ref assign_from_json.
+class code_with_referencest
+{
+public:
+  using reference_substitutiont =
+    std::function<const object_creation_referencet &(const std::string &)>;
+
+  virtual code_blockt to_code(reference_substitutiont &) const = 0;
+
+  virtual ~code_with_referencest() = default;
+};
+
+/// Code that should not contain reference
+class code_without_referencest : public code_with_referencest
+{
+public:
+  codet code;
+
+  explicit code_without_referencest(codet code) : code(std::move(code))
+  {
+  }
+
+  code_blockt to_code(reference_substitutiont &) const override
+  {
+    return code_blockt({code});
+  }
+};
+
+/// Allocation code which contains a reference.
+/// The generated code will be of the form:
+///
+///         array_length = nondet(int)
+///         assume(array_length >= 0)
+///         array_expr = new array_type[array_length];
+///
+/// Where array_length and array_expr are given by the reference substitution
+/// function.
+class reference_allocationt : public code_with_referencest
+{
+public:
+  std::string reference_id;
+  source_locationt loc;
+
+  reference_allocationt(std::string reference_id, source_locationt loc)
+    : reference_id(std::move(reference_id)), loc(std::move(loc))
+  {
+  }
+
+  code_blockt to_code(reference_substitutiont &references) const override;
+};
+
+/// Wrapper around a list of shared pointer to code_with_referencest objects,
+/// which provides a nicer interface.
+class code_with_references_listt
+{
+public:
+  std::list<std::shared_ptr<code_with_referencest>> list;
+
+  void add(code_without_referencest code);
```

Comment:
You seem to use `add(code_without_referencest{...})` a lot, so perhaps just create an `add(codet)` overload?